firedrake/assemble.py

```diff
@@ -120,15 +121,134 @@ def assemble(expr, tensor=None, bcs=None, *,
+        else:
+            visited[e] = base_form_visitor(e, tensor, bcs, diagonal,
+                                           assembly_type,
+                                           form_compiler_parameters,
+                                           mat_type, sub_mat_type,
+                                           appctx, options_prefix,
+                                           *(visited[arg] for arg in operands))
+    return visited[expr]
+
+
+def base_form_operands(expr):
+    if isinstance(expr, ufl.form.FormSum):
+        return expr.components()
+    if isinstance(expr, ufl.Adjoint):
+        return [expr.form()]
+    if isinstance(expr, ufl.Action):
+        return [expr.left(), expr.right()]
+    return []
+
+
+def base_form_visitor(expr, tensor, bcs, diagonal, assembly_type,
+                      form_compiler_parameters,
+                      mat_type, sub_mat_type,
+                      appctx, options_prefix, *args):
     if isinstance(expr, (ufl.form.Form, slate.TensorBase)):
         return assemble_form(expr, tensor, bcs, diagonal, assembly_type,
                              form_compiler_parameters,
                              mat_type, sub_mat_type,
                              appctx, options_prefix)
-    elif isinstance(expr, ufl.core.expr.Expr):
-        return assemble_expressions.assemble_expression(expr)
+    elif isinstance(expr, ufl.Adjoint):
+        if (len(args) != 1):
+            raise TypeError("Not enough operands for Adjoint")
+        mat = args[0]
+        res = PETSc.Mat().create()
+        petsc_mat = mat.M.handle
+        # TODO Add Hermitian Transpose to petsc4py and replace transpose
+        petsc_mat.transpose()
+        (row, col) = mat.arguments()
+        return matrix.AssembledMatrix((col, row), bcs, petsc_mat,
+                                      appctx=appctx,
+                                      options_prefix=options_prefix)
+    elif isinstance(expr, ufl.Action):
+        if (len(args) != 2):
+            raise TypeError("Not enough operands for Action")
+        lhs = args[0]
+        if not isinstance(lhs, matrix.MatrixBase):
+            raise TypeError("Incompatible LHS for Action")
+        rhs = args[1]
+        if isinstance(rhs, (firedrake.Cofunction, firedrake.Function)):
+            petsc_mat = lhs.M.handle
+            (row, col) = lhs.arguments()
+            res = _make_vector(col)
+
+            with rhs.dat.vec_ro as v_vec:
+                with res.dat.vec as res_vec:
+                    petsc_mat.mult(v_vec, res_vec)
+            return firedrake.Cofunction(row.function_space(), val=res.dat)
+        elif isinstance(rhs, matrix.MatrixBase):
+            petsc_mat = lhs.M.handle
+            (row, col) = lhs.arguments()
+            res = PETSc.Mat().create()
+            # TODO Figure out what goes here
+            res = petsc_mat.matMult(rhs.M.handle)
+            return matrix.AssembledMatrix(rhs.arguments(), bcs, res,
```

Comment:
It seems like we don't do anything with the `bcs` in this case, so are we "lying to children" here?